tooling/nargo_cli/src/lock.rs

```diff
@@ -0,0 +1,26 @@
+use std::{fs::File, path::PathBuf};
+
+use fs2::FileExt as _;
+
+/// A lock used to lock the Nargo.toml file so two concurrent runs of nargo
+/// commands (for example two `nargo execute`) don't overwrite output artifacts.
+pub(crate) struct Lock {
+    file: File,
+}
+
+impl Lock {
+    #[allow(clippy::self_named_constructors)]
+    pub(crate) fn lock(toml_path: PathBuf) -> Self {
+        let file = File::open(toml_path).expect("Expected Nargo.toml to exist");
+        if file.try_lock_exclusive().is_err() {
+            eprintln!("Waiting for lock on Nargo.toml...");
+        }
+
+        file.lock_exclusive().expect("Failed to lock Nargo.toml");
+        Self { file }
+    }
+
+    pub(crate) fn unlock(&self) {
```

Comment:
This looks like it could be called from `Drop` of `Lock` to make sure it happens.

Reply:
I thought about this, I'll try it. When I used it it was less clear, I think we'd still need to explicitly call `drop`, otherwise it's unlocked as soon as possible.

Reply:
I know what you mean, I am also never quite sure when Rust would drop an unused guard, however as a precaution for if it’s not unlocked I think it’s a good idea to implement Drop. If not to unlock then to panic and say it needs to be unlocked explicitly, it hasn’t been done already.

Reply:
Oh, you mean, also implement Drop. Sure, I'll do it! I wasn't exactly sure what's the semantic of Drop, if it could call anything.

Reply:
Or, I guess, if one calls `unlock` and the `drop` comes in, will it unlock it twice? I guess it's not issue, I just wasn't sure about this. And also, since now it happens in a single place in the code, maybe we don't even need the `Lock` struct, I'll check...

Reply:
Yes I suppose it would unlock twice unless you prevent it, eg with a flag, or using `ManuallyDrop` like [here](https://github.com/aakoshh/async-session-types-rs/blob/6fc5dff2d73a9c20932584fad24a553a3c3213e9/src/lib.rs#L113) 

I think Drop can do anything, but it’s difficult to do async stuff there.